### PR TITLE
661 sync logs

### LINKED ISSUE #661: Sync logs
Need to sync log table with central
* Investigate how similar logs are shaped, recorded and displayed in mSupply
* Write a proposal for translations OR
* Proposal for new table

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #661 

- Translations for activity log 
- Tests

### CODE REVIEW

File: server/service/src/sync/translations/activity_log.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+use crate::sync::sync_serde::empty_str_as_option;
+
+use super::{
+    IntegrationRecords, LegacyTableName, PullUpsertRecord, PushUpsertRecord, SyncTranslation,
+};
+
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub enum LegacyActivityLogType {
+    #[serde(rename = "user_logged_in")]
```
Comment (roxy-dao) [inline server/service/src/sync/translations/activity_log.rs:16]: Not too sure what the types will look like on mSupply 🤔

File: server/repository/migrations/postgres/2022-01-27T08-00_create_changelog_table/up.sql [hunk 2]
```diff
@@ -10,6 +10,7 @@ CREATE TYPE changelog_table_name AS ENUM (
     'stocktake_line',
     'requisition',
     'requisition_line'
+    'activity_log'
```
Comment (mark-prins) [inline server/repository/migrations/postgres/2022-01-27T08-00_create_changelog_table/up.sql:13]: for those of us who are re-using databases and wanting to test.. `ALTER TYPE changelog_table_name ADD VALUE IF NOT EXISTS 'activity_log';`

File: server/service/src/sync/test/integration/remote/activity_log.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+use serde_json::json;
+use util::{inline_edit, uuid::uuid};
+
+pub struct ActivityLogRecordTester;
+impl SyncRecordTester for ActivityLogRecordTester {
+    fn test_step_data(&self, new_site_properties: &NewSiteProperties) -> Vec<TestStepData> {
+        let mut result = Vec::new();
+        let store_id = &new_site_properties.store_id;
+
+        // STEP 1 - insert
```
Comment (mark-prins) [inline server/service/src/sync/test/integration/remote/activity_log.rs:17]: there's only one step? 🤔

File: server/service/src/sync/test/integration/remote/activity_log.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        let row = ActivityLogRow {
+            id: uuid(),
+            r#type: ActivityLogType::UserLoggedIn,
+            user_id: Some("user_account_a".to_string()),
+            store_id: None,
+            record_id: None,
+            datetime: NaiveDate::from_ymd(2020, 1, 1).and_hms(0, 0, 0),
+        };
+
+        let log_2 = inline_edit(&row, |mut l| {
```
Comment (mark-prins) [inline server/service/src/sync/test/integration/remote/activity_log.rs:28]: just curious.. where is `log_1`?

File: server/service/src/sync/test/integration/remote/mod.rs [hunk 5]
```diff
@@ -1,3 +1,4 @@
+// pub(crate) mod activity_log;
```
Comment (mark-prins) [inline server/service/src/sync/test/integration/remote/mod.rs:1]: should this and the next file have `TODO`s in them?

File: server/service/src/sync/translations/activity_log.rs (hunk 1 above)
Comment (mark-prins) [reply server/service/src/sync/translations/activity_log.rs:16]: it's a new table on mSupply is it? Not mapping to the types in the `log` table?

**Review by mark-prins — APPROVED** [review]
👍  looks good to me

File: server/service/src/sync/test/integration/remote/activity_log.rs (hunk 4 above)
Comment (roxy-dao) [reply server/service/src/sync/test/integration/remote/activity_log.rs:28]: [4c3a34b](https://github.com/openmsupply/open-msupply/pull/695/commits/4c3a34ba31ff774e04f1d6509320b84cdfe70981)

File: server/service/src/sync/test/integration/remote/mod.rs (hunk 5 above)
Comment (roxy-dao) [reply server/service/src/sync/test/integration/remote/mod.rs]: Couldn't reply to below comment for some reason. A new table has been created for the logging and the types are mapped from line 48

File: server/service/src/sync/translations/activity_log.rs (hunk 1 above)
Comment (roxy-dao) [reply server/service/src/sync/translations/activity_log.rs:16]: I mapped the type as is from omSupply

File: server/service/src/sync/test/integration/remote/activity_log.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/sync/test/integration/remote/activity_log.rs:17]: Logs won't be upserted or deleted 🤔

File: server/repository/migrations/postgres/2022-01-27T08-00_create_changelog_table/up.sql (hunk 2 above)
Comment (mark-prins) [reply server/repository/migrations/postgres/2022-01-27T08-00_create_changelog_table/up.sql:13]: I've also noticed that the `up` for `activity_log` has the column `type log_type` but the enum which is created above that is `activity_log_type`. Can you check and see which one it should be please?

Comment (roxy-dao) [reply server/repository/migrations/postgres/2022-01-27T08-00_create_changelog_table/up.sql:13]: Opps...fixed 🙏

File: server/service/src/sync/translations/activity_log.rs (hunk 1 above)
Comment (andreievg) [reply server/service/src/sync/translations/activity_log.rs:23]: > it's a new table on mSupply is it? Not mapping to the types in the log table?

Log table doesn't currently sync, should it be syncing ? (I think it's a lot of work to sync existing log table, so decided it should be new one)

**Review by andreievg — COMMENTED** [review]
@roxy-dao nice work, I've updated from 701 pr since, I think it's a good idea to chat about this PR, it would be faster (it all looks good but a couple of comments that spans multiple areas)

File: server/service/src/sync/translations/activity_log.rs (hunk 1 above)
Comment (mark-prins) [reply server/service/src/sync/translations/activity_log.rs:23]: no - shouldn't be syncing. I remember deciding that we'd use a new one. Then got myself confused when reviewing, and had a look at the types in the existing mSupply table...

File: server/service/src/sync/translations/mod.rs [hunk 6]
```diff
@@ -279,6 +284,7 @@ fn translate_changelog(
     changelog: &ChangelogRow,
 ) -> Result<Vec<RemoteSyncRecordV5>, anyhow::Error> {
     let mut translation_results = Vec::new();
+    let mut skip = false;
```
Comment (roxy-dao) [inline server/service/src/sync/translations/mod.rs:287]: Wasn't sure what to name this 🤷

File: server/service/src/sync/translations/activity_log.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            record_id,
+            datetime,
+        } = ActivityLogRowRepository::new(connection)
+            .find_one_by_id(&changelog.record_id)?
+            .ok_or(anyhow::Error::msg(format!(
+                "Activity log row ({}) not found",
+                changelog.record_id
+            )))?;
+
+        if let (Some(store_id), Some(record_id)) = (store_id, record_id) {
```
Comment (andreievg) [inline server/service/src/sync/translations/activity_log.rs:87]: Personally i have phobia with indentation, maybe it's irrational reaction to [methods like this](https://github.com/sussol/msupply/blob/master/Project/Sources/Methods/sync_process_sync_record.4dm), but i would write:
```
let (store_id, name_id, user_id) = match (store_id, name_id, user_id) {
   (Some(store_id), Some(name_id), Some(user_id)) => (store_id, name_id, user_id),
   _ => return Ok(Some(Vec::new()))
}
```

Also note above user_id in match, is there a use case for unwrap_or_default() ?

Lastly a small comment about Ok(some(Vec::new())) <- may say we want to gracefully ignore this record, `None` will result in translator not found. And it's hard to filter out prior to translator

**Review by andreievg — APPROVED** [review]
Looks good and thanks for changes, sqlite test pass but having issues with postgres. 

Approving but pending full tests pass

We probably want to merge this at the same time as: https://github.com/sussol/msupply/pull/11573, I've left a comment there to notify.

**Comment by andreievg** [comment]
Ok, postgres tests pass after this: 6e30bc6513fb8c3c4da4cece1bad2bdbb63e33fe

**Comment by Chris-Petty** [comment]
The mSupply Desktop/central server side for this is merged https://github.com/sussol/msupply/pull/11573

**Pull request merged**